### PR TITLE
[debug] ModelInspector

### LINKED ISSUE #149: extract and log grad norm for individual layers
the paper NormFormer: Improved Transformer Pretraining with Extra Normalization https://arxiv.org/abs/2110.09456 suggests that under preLN:
> gradients at earlier layers tend to be larger than gradients at later layers

so we want to verify that this is so in our case before acting on it and potentially integrating NormFormer.

So we need to expand the tensorboard and logs to log grad norm for individual layers, perhaps as in the paper we can log 5 layers: 0, 1, int(n_layers/2), -2, -1

For reference see p6 in the paper (graphs and discussion).

Currently only L2 average of all layer grad norms is calculated and logged (a single number).

------------------

Performance-wise this will require doing some extra calculations but not much - it's really just figuring out how to broadcast this info to TB so that multiple-points can be logged at once.

Additionally, we could activate this tool on demand - e.g. after encountering a spike we could roll back to the last good checkpoint and run a cycle with this debug feature enabled.

### PR DESCRIPTION
**Pull request by jaketae**
This PR resolves #149 by implementing a `ModelInspector` class similar to transformer's [`DebugUnderflowOverflow`](https://huggingface.co/transformers/debugging.html).

- Using pytorch fwd/bwd hooks log multiple things about each model's submodule and its args.

1. fwd/bwd hooks:
- [x] tensor absolute min, max
- [x] tensor absolute argmin, argmax
- [x] tensor norm  

2. optimizer
- [ ] optimizer states,  the two moving averages

3. normal TB
- [x] loss
- [x] learning rate
- [ ] gradient norm after clipping
- [ ] logits before softmax (already part of fwd hook, but mixed with all the tensors - perhaps this can have a dedicated graph

### CODE REVIEW

**Comment by jaketae** [comment]
Apologies for the lack of engagement on this issue. I'll get back on this next week after re-reading the NormFormer paper. 

As for the discussion on callback functions, I'm assuming you're suggesting that we pass in some sort of generic printing/logging function in to the tool in init, and invoke that function whenever needed. So one simplistic example would be to pass `print_rank_0` as the callback function. Alternatively, we could have it be a TB logging function. Is this the right idea?

**Comment by stas00** [comment]
I don't think this is the case. Remember this module will not have access to all data from all processes. So it'll only see a sliver of the model that each process works on. So I think each process will need to log it. And there will be a lot of data.

I think that perhaps the first step is to log from each process to a unique file, perhaps mimicking the checkpoint naming.

We will then want to somehow graph it, so perhaps let's log it in CSV format.

We need to think how to log so that this data can be graphed.

**Comment by stas00** [comment]
The original debug tool was used in just a simple 1-gpu setup, so it required no special outputs handling.

Here we are dealing with hundreds of gpus, so a simple dump to std won't do.

**Comment by stas00** [comment]
@jaketae, if that's OK with you, I will start experimenting here with how we should log things.

I am going to import `DebugUnderflowOverflow` and experiment with it first - since our main suspect is actually activations rather than grads at the moment. Once we find a good way to log, we can copy it over into the tool you have been working on.

**Comment by stas00** [comment]
This is very experimental, but this dumps all forward data into tb per rank data, which can then can be studied.

Testing with just max(tensor) for now.

snapshot_15

**Comment by jaketae** [comment]
@stas00 Thank you for making headway with the forward logger. I'll refer to your code and adapt it to the backward logger as I see fit (particularly the `tb` logic).

**Comment by stas00** [comment]
So we can't use `register_full_backward_hook` as it leads to a huge leak in deepspeed - reported it here:
https://github.com/microsoft/DeepSpeed/issues/1572
we will use the deprecated `register_backward_hook` for now.

**Comment by stas00** [comment]
If you want to play with the output, add: `--tensorboard-debug-dir debug-logs/tbs` (any path) to the options, and when it run for some dozen iterations and it exits you can view the results with:
```
tensorboard --logdir tbs
```
could probably add it to a test and see what you get.

It already does `forward` and `backward` and min/max/norm in one class.

Tomorrow going to test it on our large model.

**Comment by stas00** [comment]
Found a possibly another interesting datum to log:

Adafactor: Adaptive Learning Rates with Sublinear Memory Cost
https://arxiv.org/abs/1804.04235 

In section "5. A Problem with Adam: Out-of-Date Second Moment Estimator" it talks about RMS ratio of `(gxt)**2 / v̂xt` which is supposed to be close to 1.

logging here so it won't get lost.
